zebra-network/src/peer_set/set.rs

```diff
@@ -441,14 +424,24 @@ where
     ///
     /// Move newly ready services to the ready list if they are for peers with supported protocol
     /// versions, otherwise they are dropped. Also drop failed services.
-    fn poll_unready(&mut self, cx: &mut Context<'_>) {
-        loop {
-            match Pin::new(&mut self.unready_services).poll_next(cx) {
-                // No unready service changes, or empty unready services
-                Poll::Pending | Poll::Ready(None) => return,
+    ///
+    /// Never returns an error. If there are no unready peers, returns `Ok` immediately.
+    /// Otherwise, returns `Poll::Pending`, and registers a wakeup for the next task that becomes
+    /// ready.
+    fn poll_unready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), BoxError>> {
+        if self.unready_services.is_empty() {
+            // Don't replace the previous waker, because we have no peers to wake the task.
+            return Poll::Ready(Ok(()));
+        }
```

Comment:
This is one half of the fix: don't replace the existing waker with a waker that will never wake.